src/parse.rs

```diff
@@ -290,6 +295,68 @@ impl<'input, 'callback> Parser<'input, 'callback> {
                     }
                     self.tree[cur_ix].item.body = ItemBody::Text;
                 }
+                ItemBody::MaybeMath(can_open, _can_close, brace_context) => {
+                    let is_display = self.tree[cur_ix].next.map_or(false, |next_ix| {
+                        matches!(self.tree[next_ix].item.body, ItemBody::MaybeMath(_can_open, _can_close, _brace_context))
+                    });
+                    let mut search_count = if is_display { 2 } else { 1 };
+                    if !can_open {
+                        self.tree[cur_ix].item.body = ItemBody::Text;
+                        prev = cur;
+                        cur = self.tree[cur_ix].next;
+                        continue;
+                    }
+
+                    let result = 'result: loop {
```

Comment:
This loop should not be necessary with the unmerged `MaybeMath` approach. The search for matching inline math is handled when we come around to the second `$` (or final one of a run) in the main `handle_inline_pass1` loop.

It might also simplify things to store a bool e.g. `followed_by_dollar` in `MaybeMath`. Then you don't need to separately check the next item from the AST to verify it is display mode (other than at the last step where all items have to be erased). `MaybeEmphasis` etc use a similar idea: each char emits its own item, but also records how many of the same char follow. Up to you.

Reply:
Agreed. The idea is that if the current $ is followed by another $, inline math _cannot_ start from that point. So we recognize that, and try to find a closing _display_ tag. Only if that fails, we move to the latter $ and search for a closing inline tag.

```
$$x$
^
try to find the next $$. none found, move on

$$x$
 ^
try to find the next $. it can_close, so inline math is recognized
```

The spec test could be better, yeah.